content/FIDO2/Libraries/Using_a_library.adoc

```diff
@@ -0,0 +1,309 @@
+                    ...
+
+                    return Right.apply(
+                        new SuccessfulAuthenticationResult(
+                            request,
+                            response,
+                            userStorage.getRegistrationsByUsername(username)
+                        )
+                    );
+                ...
+                }
+
+            } else {
+                ...
+                return Left.apply(Arrays.asList("Assertion failed!", assertionTry.failed().get().getMessage()));
+            }
+    ...
+}
+----
+
+In the example above `assertRequestStorage` and `registerRequestStorage` are a link:https://en.wikipedia.org/wiki/Data_access_object[DAO] that stores
+challenges temporarily. The other DAO, `userStorage`, persists data permanently. 
+
+NOTE: FIDO2 only works on HTTPS webpages.
+
+=== Client-side  ===
+This section assumes that you are building a web site. If this is not the case,
+have a look at our link:/Software_Projects/FIDO2/FIDO2_Host_Libraries/[FIDO2 host libraries] instead.
+
+The main part of the client is to
+be a middle-man between the server and the FIDO2 compliant device.
+
+The straightforward way to use FIDO2 in a supported browser is to use the Web Authentication API, which exposes two functions:
+
+`navigator.credentials.create`:: Register using a FIDO2 device.
+`navigator.credentials.get`:: Authenticate using a FIDO2 device.
+
+==== Registration ====
+[source, html]
+----
+<script>
+if (!window.PublicKeyCredential) { /* Platform not capable. Handle error. */ }
+
+var publicKey = {
+  challenge: {challenge}, // The challenge must be produced by the server
+
+  // Relying Party:
+  rp: {
+    name: "Demo server"
+  },
+
+  // User:
+  user: {
+    id: {user_id}, // id may be generated by the server
+    name: "a.user@example.com",
+    displayName: "A User",
+    icon: "https://example.com/image.png"
+  },
+
+  // This Relying Party will accept either an ES256 or RS256 credential, but
+  // prefers an ES256 credential.
+  pubKeyCredParams: [
+    {
+      type: "public-key",
+      alg: -7 // "ES256" as registered in the IANA COSE Algorithms registry
+    },
+    {
+      type: "public-key",
+      alg: -257 // Value registered by this specification for "RS256"
+    }
+  ],
+
+  excludeCredentials: [], 
+  attestation: 'direct',
+  timeout: 60000,  
+  extensions: {"loc": true}  // Include location information in attestation
+};
+
+// Note: The following call will cause the authenticator to display UI.
+navigator.credentials.create({ publicKey })
+  .then(function (attestation) {
+    // Send new credential info to server for verification and registration.
+  }).catch(function (err) {
+    // No acceptable authenticator or user refused consent. Handle appropriately.
+  });
+</script>
+----
+
+==== Authentication ====
+[source, html]
+----
+<script>
+if (!window.PublicKeyCredential) { /* Platform not capable. Handle error. */ }
+
+  navigator.credentials.get({{
```

Comment:
Why are there double `{{`s and `}}`s throughout this object literal?

Reply:
Thanks for catching those copy-paste-errors.